[iOS] PointerLockOptions.unadjustedMovement is supported when a pointing device is available

aprotyas opened the pull request:
#### a515900071861d7aebe336fddcae0ee204b5e5e8
<pre>
[iOS] PointerLockOptions.unadjustedMovement is supported when a pointing device is available
<a href="https://bugs.webkit.org/show_bug.cgi?id=297304">https://bugs.webkit.org/show_bug.cgi?id=297304</a>
<a href="https://rdar.apple.com/158184340">rdar://158184340</a>

Reviewed by Aditya Keerthi.

Given we will be generating pointermove deltas from unaccelerated data
(c.f. webkit.org/b/297305) to begin with, this option is unconditionally
supported on iOS family whenever a pointing device is available.

This patch exposes WebPage::hasAccessoryMousePointingDevice(), which
keys off of the WebProcess singleton, allowing PointerLockController to
evaluate whether the `unadjustedMovement` option can be respected. Note
that this only meaningfully changes the iOS behavior because we
unconditionally report that a mouse pointing device is present on other
platforms.

Moreover, this patch improves some iOS test coverage of the pointer lock
feature through pointer-lock-option-unadjusted-movement.html. To get
this test running, we need to force toggle the PointerLockEnabled
preference. We also need to mock the web process into thinking a mouse
pointing device is available, which we accomplish through new TestRunner
interface.

* LayoutTests/http/tests/resources/pointer-lock/pointer-lock-test-harness.js:
(runWithUserGesture):
(runOnKeyPress.keypressHandler): Deleted.
(runOnKeyPress): Deleted.
(doNextStep): Deleted.

Instead of emulating the act of obtaining user gesture with a keypress,
let us just run the test steps with user gesture granted (through our
Internals testing interface).

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/pointer-lock/pointer-lock-option-unadjusted-movement-expected.txt: Added.
* LayoutTests/pointer-lock/pointer-lock-option-unadjusted-movement.html:

Altered the test slightly to obtain keyWindow/FR status before running
the test steps, since this is a prerequisite for pointer lock
activation.

* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/PointerLockController.cpp:
(WebCore::PointerLockController::rejectPromises):
(WebCore::PointerLockController::supportsUnadjustedMovement): Deleted.
* Source/WebCore/page/PointerLockController.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::hasAccessoryMousePointingDevice const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::mediaSessionManager const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::isMousePrimaryPointingDevice):
(WebKit::hasAccessoryMousePointingDevice): Deleted.
* Source/WebKitLegacy/ios/WebCoreSupport/WebChromeClientIOS.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setHasMouseDeviceForTesting):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setHasMouseDeviceForTesting):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::platformResetStateToConsistentValues):
(WTR::TestController::setHasMouseDeviceForTesting):

Keep hold of the shared WKMouseDeviceObserver instance and start/stop as
appropriate.

Canonical link: <a href="https://commits.webkit.org/298635@main">https://commits.webkit.org/298635@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3fad3cf73bc0bebadf5e629f702f94970fab7990

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116087 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35748 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26291 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122143 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66638 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/fbbe79e4-1c15-4e59-8715-8b50cdbf3f6a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117976 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36442 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44336 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88192 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6071fd4a-f30f-4a6a-a124-bcad0cd90b68) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119035 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29070 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104164 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68603 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28186 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22271 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65825 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98458 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22409 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125293 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42981 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32265 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96926 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43346 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100354 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96710 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41982 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19868 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38927 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18552 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42868 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48460 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42335 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45670 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44039 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder reviewed (approved):
> Morevoer

In the commit message.

pxlcoder commented on Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
```diff
@@ -619,4 +634,20 @@ static _WKFocusStartsInputSessionPolicy focusStartsInputSessionPolicy(const Test
+#if HAVE(MOUSE_DEVICE_OBSERVATION)
+    if (!m_mouseDeviceObserver) {
+        m_mouseDeviceObserver = [NSClassFromString(@"WKMouseDeviceObserver") sharedInstance];
+        [m_mouseDeviceObserver start];
```
Slightly surprised the explicit `start` is necessary, since it should be called by `WebProcessProxy::platformInitialize`?

aprotyas replied:
That's true, I was following prior art from TestWebKitAPI (maybe it's unnecessary there too)

aprotyas replied:
I suppose the test will run correctly without this explicit `start` call. Let me try it out.

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/069a59899a7a135ed25ed882b5154e9880b4a9b1

pxlcoder commented on Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
```diff
@@ -458,6 +469,10 @@ static _WKFocusStartsInputSessionPolicy focusStartsInputSessionPolicy(const Test
 
+#if HAVE(MOUSE_DEVICE_OBSERVATION)
+    [std::exchange(m_mouseDeviceObserver, nil) stop];
+#endif
```
There's an early return on line 465, so this should be moved up next to other the state resetting.

pxlcoder replied:
(and same general question about the explicit start/stop)

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/069a59899a7a135ed25ed882b5154e9880b4a9b1

aprotyas committed:
[iOS] PointerLockOptions.unadjustedMovement is supported when a pointing device is available
https://bugs.webkit.org/show_bug.cgi?id=297304
rdar://158184340

Reviewed by Aditya Keerthi.

Given we will be generating pointermove deltas from unaccelerated data
(c.f. webkit.org/b/297305) to begin with, this option is unconditionally
supported on iOS family whenever a pointing device is available.

This patch exposes WebPage::hasAccessoryMousePointingDevice(), which
keys off of the WebProcess singleton, allowing PointerLockController to
evaluate whether the `unadjustedMovement` option can be respected. Note
that this only meaningfully changes the iOS behavior because we
unconditionally report that a mouse pointing device is present on other
platforms.

Moreover, this patch improves some iOS test coverage of the pointer lock
feature through pointer-lock-option-unadjusted-movement.html. To get
this test running, we need to force toggle the PointerLockEnabled
preference. We also need to mock the web process into thinking a mouse
pointing device is available, which we accomplish through new TestRunner
interface.

* LayoutTests/http/tests/resources/pointer-lock/pointer-lock-test-harness.js:
(runWithUserGesture):
(runOnKeyPress.keypressHandler): Deleted.
(runOnKeyPress): Deleted.
(doNextStep): Deleted.

Instead of emulating the act of obtaining user gesture with a keypress,
let us just run the test steps with user gesture granted (through our
Internals testing interface).

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/pointer-lock/pointer-lock-option-unadjusted-movement-expected.txt: Added.
* LayoutTests/pointer-lock/pointer-lock-option-unadjusted-movement.html:

Altered the test slightly to obtain keyWindow/FR status before running
the test steps, since this is a prerequisite for pointer lock
activation.

* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/PointerLockController.cpp:
(WebCore::PointerLockController::rejectPromises):
(WebCore::PointerLockController::supportsUnadjustedMovement): Deleted.
* Source/WebCore/page/PointerLockController.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::hasAccessoryMousePointingDevice const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::mediaSessionManager const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::isMousePrimaryPointingDevice):
(WebKit::hasAccessoryMousePointingDevice): Deleted.
* Source/WebKitLegacy/ios/WebCoreSupport/WebChromeClientIOS.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setHasMouseDeviceForTesting):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setHasMouseDeviceForTesting):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::platformResetStateToConsistentValues):
(WTR::TestController::setHasMouseDeviceForTesting):

Keep hold of the shared WKMouseDeviceObserver instance and start/stop as
appropriate.

Canonical link: https://commits.webkit.org/298635@main
